### PR TITLE
feat: create a custom camera 

### PR DESCRIPTION
**Pull request by ice-alcides**
## Key Changes

Figma [link](https://www.figma.com/design/3TnZngoFeklOVC8TexlR1D/%F0%9F%A7%8A-ICE-Cloud?node-id=38168-132546&node-type=text&t=8V43okZy8VZdCBkO-0) 

## Context
- **First PR in Stories Flow**: This is the initial pull request in a series that will implement the complete Stories feature.
- **Foundation for Future Work**: The components and structure introduced here will serve as the foundation for subsequent PRs in this feature implementation.

### 1. **StoryCameraController for Camera State Management**
   - Developed a `StoryCameraController` provider to manage video recording and flash operations.
   - Implemented automatic gallery saving for recorded videos.

### 2. **useRecordingController Hook**
   - Created a custom hook to manage recording state and progress.
   - Implemented automatic recording stop at maximum duration.

### 3. **Camera UI Components**
   - Added `IdleCameraPreview` for non-recording states, including control buttons for various camera actions.
   - Developed `StoryCameraPreview` for aspect-ratio-adjusted camera preview.

### 4. **Recording Interaction**
   - Introduced `CaptureButton` with a long-press gesture for intuitive video recording.
   - Implemented `CircularRecordingIndicator` for visual feedback during recording.

## TODO
- [ ] Implement app lifecycle state management to handle app backgrounding during recording properly.
- [ ] Implement gallery opening functionality

### CODE REVIEW

File: lib/app/features/feed/create_story/views/components/circular_recording_indicator.dart [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return SizedBox.square(
+      dimension: 90.0.s,
+      child: DecoratedBox(
+        decoration: const BoxDecoration(
+          shape: BoxShape.circle,
+          color: Colors.transparent,
+        ),
+        child: CircularProgressIndicator(
```
[the remark this answers is not in the document]
Comment (ice-alcides) [reply lib/app/features/feed/create_story/views/components/circular_recording_indicator.dart:23]: Yes, this indicator is circular (with a blue color), as in the screenshot:

CleanShot 2024-10-11 at 11 15 41@2x

File: lib/app/features/feed/create_story/views/components/inner_capture_circle.dart [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  const InnerCaptureCircle({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return SizedBox.square(
+      dimension: 54.0.s,
+      child: const DecoratedBox(
+        decoration: BoxDecoration(
+          shape: BoxShape.circle,
+          color: Colors.white,
```
[the remark this answers is not in the document]
Comment (ice-alcides) [reply lib/app/features/feed/create_story/views/components/inner_capture_circle.dart:16]: thanks, I will replace

File: lib/app/features/feed/create_story/views/pages/story_camera_widget.dart [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    final cameraControllerAsync = ref.watch(cameraControllerNotifierProvider);
+    final storyCameraState = ref.watch(storyCameraControllerProvider);
+    final storyCameraNotifier = ref.read(storyCameraControllerProvider.notifier);
+
+    final (recordingDuration, recordingProgress) = useRecordingController(
+      ref,
+      isRecording: storyCameraState.isRecording,
+    );
+
+    Future<void> startRecording() async => storyCameraNotifier.startVideoRecording();
```
Comment (ice-tychon) [inline lib/app/features/feed/create_story/views/pages/story_camera_widget.dart:31]: nit: wrap with `useCallback`

**Review by ice-tychon — APPROVED** [review]

File: lib/app/features/feed/create_story/views/components/inner_capture_circle.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (ice-alcides) [reply lib/app/features/feed/create_story/views/components/inner_capture_circle.dart]: done

File: lib/app/features/feed/create_story/views/components/recording_indicator.dart [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import 'package:ice/generated/assets.gen.dart';
+
+class RecordingIndicator extends StatelessWidget {
+  const RecordingIndicator({
+    required this.recordingDuration,
+    super.key,
+  });
+  final Duration recordingDuration;
+
+  String formatDuration(Duration duration) {
```
[the remark this answers is not in the document]
Comment (ice-alcides) [reply lib/app/features/feed/create_story/views/components/recording_indicator.dart]: done

File: lib/app/features/feed/create_story/create_story_modal.dart [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+          ),
+          const Spacer(),
+          Center(
+            child: PermissionAwareWidget(
+              permissionType: Permission.camera,
+              builder: (context, onPressed) {
+                return GestureDetector(
+                  onTap: onPressed,
+                  child: const Text(
+                    'Create Story',
```
[the remark this answers is not in the document]
Comment (ice-alcides) [reply lib/app/features/feed/create_story/create_story_modal.dart]: removed

File: lib/app/features/feed/create_story/views/components/story_camera_preview.dart [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  @override
+  Widget build(BuildContext context) {
+    if (!controller.value.isInitialized) {
+      return const Center(child: IceLoadingIndicator());
+    }
+
+    final size = MediaQuery.sizeOf(context);
+    final cameraAspectRatio = controller.value.aspectRatio;
+
+    return ClipRRect(
```
[the remark this answers is not in the document]
Comment (ice-alcides) [reply lib/app/features/feed/create_story/views/components/story_camera_preview.dart:25]: We should show rounded corners around the custom camera by design in Figma. These rounded edges are not considered in the recorded video, but we can ignore them as they are hardly noticeable.

File: lib/app/features/feed/create_story/views/pages/story_camera_widget.dart (hunk 3 above)
Comment (ice-alcides) [reply lib/app/features/feed/create_story/views/pages/story_camera_widget.dart]: done

**Review by ice-tychon — APPROVED** [review]

**Review by ice-kreios — APPROVED** [review]

**Pull request merged**